atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/cache/TransactionScopedCacheImplTest.java

```diff
@@ -138,6 +152,186 @@ public void allValuesReadFromCachePreventsReadToDb() {
+        ValueCacheSnapshot snapshot = ValueCacheSnapshotImpl.of(
+                HashMap.of(
+                        CellReference.of(TABLE, createCell(1, 1)),
+                        CacheEntry.unlocked(CacheValue.empty()),
+                        CellReference.of(TABLE, createCell(1, 2)),
+                        CacheEntry.unlocked(CacheValue.empty()),
+                        CellReference.of(TABLE, createCell(1, 3)),
+                        CacheEntry.unlocked(CacheValue.empty()),
+                        CellReference.of(TABLE, createCell(2, 1)),
+                        CacheEntry.unlocked(CacheValue.empty()),
+                        CellReference.of(TABLE, createCell(2, 3)),
+                        CacheEntry.unlocked(CacheValue.empty())),
+                HashSet.of(TABLE),
+                ImmutableSet.of(TABLE));
+        TransactionScopedCache cache = TransactionScopedCacheImpl.create(snapshot, metrics);
+
+        ImmutableList<byte[]> rowsAndCols = ImmutableList.of(createBytes(1), createBytes(2), createBytes(3));
+        ColumnSelection columns = ColumnSelection.create(rowsAndCols);
+
+        Multiset<Cell> directLookups = HashMultiset.create();
+        NavigableSet<byte[]> rowLookups = new TreeSet<>(UnsignedBytes.lexicographicalComparator());
+
+        NavigableMap<byte[], RowResult<byte[]>> lookup =
+                cache.getRows(TABLE, rowsAndCols, columns, cells -> loadCells(cells, directLookups), rows -> {
+                    rows.forEach(rowLookups::add);
+                    return new TreeMap<>(UnsignedBytes.lexicographicalComparator());
+                });
+
+        Cell onlyCell = createCell(2, 2);
+        assertThat(directLookups).containsExactly(onlyCell);
+        assertThat(rowLookups).containsExactly(createBytes(3));
+        assertThat(lookup.size()).isEqualTo(1);
+        assertThat(lookup.get(createBytes(2)).getOnlyColumnValue()).containsExactly(createBytes(onlyCell));
+    }
+
+    @Test
+    public void getRowsDoesNotBreakOnEmptyRowsAndCachesThemCorrectly() {
+        TransactionScopedCache cache = TransactionScopedCacheImpl.create(snapshotWithSingleValue(), metrics);
+
+        ImmutableList<byte[]> rowsAndCols = ImmutableList.of(createBytes(2), createBytes(3));
+        ColumnSelection columns = ColumnSelection.create(rowsAndCols);
+
+        Multiset<Cell> directLookups = HashMultiset.create();
+
+        AtomicInteger rowLookups = new AtomicInteger();
+        Function<Iterable<byte[]>, NavigableMap<byte[], RowResult<byte[]>>> rowLookup = rows -> {
+            rowLookups.incrementAndGet();
+            return new TreeMap<>(UnsignedBytes.lexicographicalComparator());
+        };
+        NavigableMap<byte[], RowResult<byte[]>> lookup =
+                cache.getRows(TABLE, rowsAndCols, columns, cells -> loadCells(cells, directLookups), rowLookup);
+        cache.getRows(TABLE, rowsAndCols, columns, cells -> loadCells(cells, directLookups), rowLookup);
+
```

Comment:
we should add `assertThatCode(cache.getRows...).doesNotThrowAnyException()` if that is your intention

Reply:
This was more to see that the second call does no remote lookups, not necessarily that it does not throw